osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -274,13 +273,17 @@ private void updateSearch()
                                                        Filter.DisplayStyleControl.Dropdown.Current.Value,
                                                        Filter.Tabs.Current.Value); //todo: sort direction (?)
 
-            getSetsRequest.Success += r =>
+            getSetsRequest.Success += response =>
             {
-                BeatmapSets = r?.
-                                Select(response => response.ToBeatmapSet(rulesets)).
-                                Where(b => beatmaps.QueryBeatmapSet(q => q.OnlineBeatmapSetID == b.OnlineBeatmapSetID) == null);
+                Task.Run(() =>
+                {
+                    var onlineIds = response.Select(r => r.OnlineBeatmapSetID).ToList();
+                    var presentOnlineIds = beatmaps.QueryBeatmapSets(s => onlineIds.Contains(s.OnlineBeatmapSetID)).Select(r => r.OnlineBeatmapSetID).ToList();

```

Comment:
How about making this into a hashet? `new HashSet(IEnumerable)`

Reply:
Two things:
- I don't think that will help because the full list content is being passed to SQLite anyway.
- I'm not sure `IQueryable` will handle the `HashSet` correctly (though it probably will), but see point 1

Reply:
Actually can't you just Intersect the two previous results? Intersect the database ids with the response ids.